src/common/internal/test_group.ts

```diff
@@ -428,40 +476,83 @@ class RunCaseSpecific implements RunCase {
       return didSeeFail ? 'fail' : 'pass';
     };
 
-    rec.start();
-    if (this.subcases) {
-      let totalCount = 0;
-      let skipCount = 0;
-      for (const subParams of this.subcases) {
-        rec.info(new Error('subcase: ' + stringifyPublicParams(subParams)));
-        try {
-          const params = mergeParams(this.params, subParams);
-          const subcaseQuery = new TestQuerySingleCase(
-            selfQuery.suite,
-            selfQuery.filePathParts,
-            selfQuery.testPathParts,
-            params
-          );
-          await this.runTest(rec, params, true, getExpectedStatus(subcaseQuery));
-        } catch (ex) {
-          if (ex instanceof SkipTestCase) {
-            // Convert SkipTestCase to info messages
-            ex.message = 'subcase skipped: ' + ex.message;
-            rec.info(ex);
-            ++skipCount;
-          } else {
-            // Since we are catching all error inside runTest(), this should never happen
-            rec.threw(ex);
+    try {
+      rec.start();
+      const sharedState = this.fixture.MakeSharedState(this.params);
+      try {
+        await sharedState.doInit();
+        if (this.beforeFn) {
+          await this.beforeFn(sharedState);
+        }
+
+        const finalizePromises: Promise<void>[] = [];
+        if (this.subcases) {
+          let totalCount = 0;
+          let skipCount = 0;
+          for (const subParams of this.subcases) {
+            rec.info(new Error('subcase: ' + stringifyPublicParams(subParams)));
+            try {
+              const params = mergeParams(this.params, subParams);
+              const subcaseQuery = new TestQuerySingleCase(
+                selfQuery.suite,
+                selfQuery.filePathParts,
+                selfQuery.testPathParts,
+                params
+              );
+              await this.runTest(
+                rec,
+                sharedState,
+                params,
+                /* throwSkip */ true,
+                getExpectedStatus(subcaseQuery),
+                finalizePromises
+              );
+            } catch (ex) {
+              if (ex instanceof SkipTestCase) {
+                // Convert SkipTestCase to info messages
+                ex.message = 'subcase skipped: ' + ex.message;
+                rec.info(ex);
+                ++skipCount;
+              } else {
+                // Since we are catching all error inside runTest(), this should never happen
+                rec.threw(ex);
+              }
+            }
+            ++totalCount;
+
+            // Wait to ensure there aren't too many in-flight subcases.
+            const kMaxPendingEventualExpectations = 500;
+            while (sharedState.numPendingEventualExpectations > kMaxPendingEventualExpectations) {
+              await sharedState.makeEventualExpectationDecrementedPromisePromise();
+            }
```

Comment:
Can we just do a maximum number of pending subcases instead? Even just batch them into groups of 500 (run 500 in parallel, wait for them to finalize, run another 500, and so on). Seems a bit simpler than counting EventualExpectations.